### PR TITLE
[nextercism] Create solution metadata

### PR DESCRIPTION
**Pull request by kytrinyx**
We are going to store some metadata about a solution in a hidden file.
This currently only hides the file on Linux/MacOS.
I need to work out the hidden file for Windows, presumably using
`syscall.SetFileAttributes(f, syscall.FILE_ATTRIBUTE_HIDDEN)`.

We're going to use this in the `download` (store the metadata), `submit` (find the solution id), and `open` (look up the web url) commands.

### CODE REVIEW

File: workspace/solution.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+// Solution contains metadata about a user's solution.
+type Solution struct {
+	Track       string    `json:"track"`
+	Exercise    string    `json:"exercise"`
+	ID          string    `json:"id"`
+	URL         string    `json:"url"`
+	Handle      string    `json:"handle"`
+	IsRequester bool      `json:"is_requester"`
+	SubmittedAt time.Time `json:"submitted_at"`
```
Comment (kytrinyx) [inline workspace/solution.go]: I just realized that we will sometimes download a fresh exercise where the solution hasn't been submitted yet. There will still be a _solution_. It has an ID, URL, etc. It just hasn't been submitted, so there's no timestamp.

I think I need a custom time format that can be a pointer.

Comment (kytrinyx) [inline workspace/solution.go]: Ok, I made time.Time a pointer, even though we have to jump through some hoops to take an address from a time.Time value. We'll see how painful it is and can always change it later if necessary.

**Comment by kytrinyx** [comment]
OK, super not sure about the windows hidden file thing. I've never tried this before. I read the docs and the source code for the `syscall` stuff. I can't test this locally (no Windows), so I kind of expect to have gotten this wrong.

If someone would pull this down on Windows and try running the tests (wait, how do you do `go test $(go list ./... | grep -v vendor)` on windows?) that would be a great help.

If you have the spare cycles to do something like the below in a temporary folder and see what happens, that would be cool too:

```
package main

import (
	"io/ioutil"
	"os"

	"github.com/exercism/cli/visibility"
)

func main() {
	f := `can-you-see-me.txt`
	ioutil.WriteFile(f, []byte(`o hai, didn't see you there`), os.FileMode(0755))
	visibility.HideFile(f)
}
```

Then figure out if `can-you-see-me.txt` is hidden or not (I don't even know how you would find out, sorry!).

**Comment by Tonkpils** [comment]
Does this help? https://github.com/inconshreveable/go-update/blob/master/hide_windows.go I'm not on a windows computer but I know that works as I've tested it on it before.

**Comment by kytrinyx** [comment]
Oh, yeah. Interesting. I would definitely not have thought about that approach. Did you copy/paste to use it? I guess you'd have to since it's not exported. I wish they'd export it, since we're already relying on that package. We could just call it directly.

**Comment by robphoenix** [comment]
I'll give this a go tomorrow on my Windows laptop

**Comment by kytrinyx** [comment]
I added a few more commits to this one. Realized that we're going to want to display solutions to people to get them to disambiguate. E.g. when they've got three different "clock" exercises, one in Go, one in Rust, and one by some other user, if they say `exercism open clock`, we need to get them to pick. So I've fleshed out solution a little bit to give us a `String()` method that will help us with that.

File: visibility/hide_file_windows.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	ptr, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return err
+	}
+	attributes, err := syscall.GetFileAttributes(ptr)
+	if err != nil {
+		return err
+	}
+
+	attrs |= syscall.FILE_ATTRIBUTE_HIDDEN
```
Comment (robphoenix) [inline visibility/hide_file_windows.go]: s/attrs/attributes

Comment (kytrinyx) [inline visibility/hide_file_windows.go]: whuuut. This is what I get when I can't lean on the compiler!

**Comment by kytrinyx** [comment]
> I'm not sure if it's me and not knowing the right build instructions, but I had to comment out the hide_file.go contents as it was complaining about the function being redeclared. I suppose that won't be a problem with a binary will it.

I will go look at instructions about build constraints to see if I got it wrong. It shouldn't give you trouble if I do it right, I think.

**Comment by kytrinyx** [comment]
@robphoenix You're right. This built just fine when I just built the binary. I don't understand why it's complaining on your machine.

I'm going to go ahead and merge this, as I really need to get on with the various commands that depend on this one, but I would love to figure the build thing out before too long.

**Comment by robphoenix** [comment]
Cool, I'll have a look into it too, see what I can figure out.

**Pull request merged**